Feat/mainnet test

Seomingyun opened the pull request:
## Summary
<!-- 어떤 내용을 개발했는지 요약해서 설명하고, 관련된 이슈 멘션 -->
emergency withdraw function 추가
이후 code size limit 때문에 전체적으로 code size를 줄이기 위해 로직 수정

**Issue Number**
N/A

**PR Type**
<!-- Please check the one that applies to this PR. -->

- [x] Bugfix
- [x] Feature
- [ ] Design
- [ ] Test code
- [ ] Other... Please describe:

## Description 
<!-- 어떤 내용을 개발했는지 자세히 설명하고 싶은 것들 설명 -->

1. emergencyWithdraw function 추가 + 이에 따른 test code 추가
2. code size limit을 위한 custom error 사용
3. code size limit을 위한 withdraw logic 변경
4. code size limit을 위한 rebalance logic 변경
5. code size limit을 위한 SwapModule의 getPair function 삭제 및 변경

## Code Review Checklist
<!-- 어떤 내용을 중점적으로 code review 해주면 좋겠는지 작성 -->
- [ ] withdraw logic
- [ ] rebalance logic
- [ ] swapModule의 getPair 로직


## Other information
관련 노션
code size limit 관련: 
https://www.notion.so/quinoaprotocol/Product-contract-code-size-656cef9b8aef4aeabfeb8ace92204d91
custom error 관련: 
https://www.notion.so/quinoaprotocol/solidity-custom-error-b88e622a60bb495b8e764675fdd8fac0

Seomingyun committed:
add emregencyWithdraw function for main network test

Seomingyun committed:
change function name(emergencyWithdrawForTest→emergencyWithdraw)

Seomingyun committed:
add emergencyWithdraw event

Seomingyun committed:
add get function in ISwapModule

Seomingyun committed:
don't transfer when thransferAmount is 0

Seomingyun committed:
delete getPair logic for reducing code size

Seomingyun committed:
add prefix _ for internal function

Seomingyun committed:
fix logic for reducing code size

Seomingyun committed:
fix test logic & delete unused comments

Seomingyun committed:
modify optimizer runs option

wakii commented on contracts/Product.sol:
```diff
@@ -395,6 +425,21 @@ contract Product is ERC20, IProduct, SwapModule, AutomationCompatibleInterface {
     }
 
+    function emergencyWithdraw() external onlyDac {
+        if(isActive) revert DisabledNow(isActive);
```
이거 호출하면 deactive 하게 하는게 낫지 않을까요?

Seomingyun replied:
test 용으로만 남겨놓는거 맞져 ?? 그럼 상관없을듯요

wakii replied:
넹~

wakii reviewed (commented):
withdraw는 한 번 설명 부탁드려도 되나요?? 혼자 보려니까 오래걸리네요

Seomingyun committed:
modify Deposit/Withdraw event paramter(add share price and time stamp)

Seomingyun commented:
> withdraw는 한 번 설명 부탁드려도 되나요?? 혼자 보려니까 오래걸리네요

https://www.notion.so/quinoaprotocol/Product-contract-code-size-656cef9b8aef4aeabfeb8ace92204d91
설명 추가했습니당

Seomingyun committed:
change emergencyWithdraw logic

when dac call emergencyWithdraw function, the product is automatically changed to <deactivation status>

Seomingyun committed:
delete unused comments

Seomingyun committed:
fix deactivate logic in emergencyWithdraw function

Seomingyun committed:
add event logging
